File: src/main/javabak/multiblock/MultiblockBlock.java
package redcrafter07.processed.multiblock;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.block.state.BlockState;
import redcrafter07.processed.block.machine_abstractions.ProcessedBlock;

public abstract class MultiblockBlock extends ProcessedBlock {
    public MultiblockBlock(Properties properties) {
        super(properties);
    }

    @Override
    protected BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor level, BlockPos pos, BlockPos neighborPos) {
        if (!(neighborState.getBlock() instanceof CasingBlock) && !pos.equals(neighborPos)) {
            if (level.getBlockEntity(pos) instanceof MultiblockBlockEntity be) be.recheck();
        }

        return super.updateShape(state, direction, neighborState, level, pos, neighborPos);
    }

    @Override
    public void onRemove(BlockState state, Level level, BlockPos pos, BlockState newState, boolean isMoving) {
        if (level.getBlockEntity(pos) instanceof MultiblockBlockEntity be) be.onRemove(level);
        super.onRemove(state, level, pos, newState, isMoving);
    }
}

File: src/main/kotlin/redcrafter07/processed/multiblock/MultiblockBlock.kt
package redcrafter07.processed.multiblock

import net.minecraft.core.BlockPos
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.BlockHitResult
import redcrafter07.processed.block.machine_abstractions.ProcessedBlock

abstract class MultiblockBlock(properties: Properties) : ProcessedBlock(properties) {
    public override fun onRemove(
        state: BlockState, level: Level, pos: BlockPos, newState: BlockState, movedByPiston: Boolean
    ) {
        if(newState.`is`(state.block)) return
        MultiblockPreview.removeIfLastDisplayed(pos)
        val be = level.getBlockEntity(pos)
        if (be is MultiblockBlockEntity) be.onRemove(level)
        super.onRemove(state, level, pos, newState, movedByPiston)
    }

    override fun useWithoutItem(
        state: BlockState, level: Level, pos: BlockPos, player: Player, hitResult: BlockHitResult
    ): InteractionResult {
        if (player.isShiftKeyDown) {
            val be = level.getBlockEntity(pos)
            if (be is MultiblockBlockEntity) {
                if (level.isClientSide) MultiblockPreview.update(be)
                return InteractionResult.sidedSuccess(level.isClientSide)
            }
        }
        return super.useWithoutItem(state, level, pos, player, hitResult)
    }
}
